cpu/samd21/cpu.c

```diff
@@ -72,19 +94,35 @@ static void clk_init(void)
     while (!(SYSCTRL->PCLKSR.reg & SYSCTRL_PCLKSR_OSC8MRDY)) {}
 #endif
 
+    /* Setup GCLK2 with divider 1 (32.768kHz) */
+    GCLK->GENDIV.reg  = (GCLK_GENDIV_ID(SAM0_GCLK_32KHZ)  | GCLK_GENDIV_DIV(0));
+    GCLK->GENCTRL.reg = (GCLK_GENCTRL_ID(SAM0_GCLK_32KHZ) | GCLK_GENCTRL_GENEN
+                      | GCLK_GENCTRL_RUNSTDBY
+#if GEN2_ULP32K
+                      | GCLK_GENCTRL_SRC_OSCULP32K);
+#else
+                      | GCLK_GENCTRL_SRC_XOSC32K);
+
+    SYSCTRL->XOSC32K.reg = SYSCTRL_XOSC32K_ONDEMAND
+                         | SYSCTRL_XOSC32K_EN32K
+                         | SYSCTRL_XOSC32K_XTALEN
+                         | SYSCTRL_XOSC32K_STARTUP(6)
+                         | SYSCTRL_XOSC32K_ENABLE;
+#endif
+
 #if CLOCK_USE_PLL
     /* reset the GCLK module so it is in a known state */
     GCLK->CTRL.reg = GCLK_CTRL_SWRST;
     while (GCLK->STATUS.reg & GCLK_STATUS_SYNCBUSY) {}
 
     /* setup generic clock 1 to feed DPLL with 1MHz */
     GCLK->GENDIV.reg = (GCLK_GENDIV_DIV(8) |
-                        GCLK_GENDIV_ID(1));
+                        GCLK_GENDIV_ID(SAM0_GCLK_1MHZ));
     GCLK->GENCTRL.reg = (GCLK_GENCTRL_GENEN |
                          GCLK_GENCTRL_SRC_OSC8M |
-                         GCLK_GENCTRL_ID(1));
-    GCLK->CLKCTRL.reg = (GCLK_CLKCTRL_GEN(1) |
-                         GCLK_CLKCTRL_ID(1) |
+                         GCLK_GENCTRL_ID(SAM0_GCLK_1MHZ));
+    GCLK->CLKCTRL.reg = (GCLK_CLKCTRL_GEN(SAM0_GCLK_1MHZ) |
+                         GCLK_CLKCTRL_ID(SAM0_GCLK_1MHZ)  |
                          GCLK_CLKCTRL_CLKEN);
```

Comment:
Let me apologize upfront, because I am constantly confused about clock generators. So I may talk nonsense.
I'm afraid that `GCLK_CLKCTRL_ID(SAM0_GCLK_1MHZ)` is not correct. The `1` in this case is `DPLL`.
And `GCLK_GENCTRL_ID(SAM0_GCLK_1MHZ)` is probably not correct either. The `1` in this case is `GCLKIN`. No?

Reply:
Yea let's not use `GCLK_CLKCTRL_ID()` when more explicit defines are available instead.
`GCLK_GENCTRL_ID(SAM0_GCLK_1MHZ)` should point to `GCLKGEN1` which is fine.

Reply:
OK